### PR TITLE
Deploying a MVP BinderHub

### PR DESCRIPTION
**Pull request by sgibson91**
related #1280 

This PR deploys a minimal working BinderHub to the 2i2c cluster. It can be accessed at https://binder-staging.2i2c.cloud

* Basing Binderhub's values from the original Pangeo Binder's values: https://github.com/pangeo-data/pangeo-binder/blob/staging/pangeo-binder/values.yaml
* Config for the BinderHub has been added
  * For now the BinderHub pushes images to my personal Docker Hub account
* Deployer has been updated to handle a BinderHub deployment
* All extra config entries have been disabled in the BinderHub helm chart for now, until we migrate over the templates from basehub

### CODE REVIEW

**Comment by sgibson91** [comment]
The above looks like the following PR but we have not enabled the prePuller or userPlaceholders

* https://github.com/2i2c-org/infrastructure/pull/1390

**Comment by sgibson91** [comment]
From the hub logs

```bash
k logs -c hub hub-66958dc964-mtd6s --previous
Loading /usr/local/etc/jupyterhub/secret/values.yaml
No config at /usr/local/etc/jupyterhub/existing-secret/values.yaml
Loading extra config: 0-binderspawnermixin
Loading extra config: 00-binder
Loading extra config: 01-custom-theme
Loading extra config: 02-custom-admin
[E 2022-06-08 16:01:34.455 JupyterHub app:2969]
    Traceback (most recent call last):
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/app.py", line 2966, in launch_instance_async
        await self.initialize(argv)
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/app.py", line 2457, in initialize
        self.load_config_file(self.config_file)
      File "/usr/local/lib/python3.8/dist-packages/traitlets/config/application.py", line 87, in inner
        return method(app, *args, **kwargs)
      File "/usr/local/lib/python3.8/dist-packages/traitlets/config/application.py", line 775, in load_config_file
        for (config, filename) in self._load_config_files(filename, path=path, log=self.log,
      File "/usr/local/lib/python3.8/dist-packages/traitlets/config/application.py", line 737, in _load_config_files
        config = loader.load_config()
      File "/usr/local/lib/python3.8/dist-packages/traitlets/config/loader.py", line 616, in load_config
        self._read_file_as_dict()
      File "/usr/local/lib/python3.8/dist-packages/traitlets/config/loader.py", line 648, in _read_file_as_dict
        exec(compile(f.read(), conf_filename, 'exec'), namespace, namespace)
      File "/usr/local/etc/jupyterhub/jupyterhub_config.py", line 446, in <module>
        exec(config_py)
      File "<string>", line 3, in <module>
    ModuleNotFoundError: No module named 'jupyterhub_configurator'
```

**Comment by sgibson91** [comment]
Commenting out all the custom scripts made the hub pod get out of crashloopbackoff

**Comment by sgibson91** [comment]
Binder exists at https://binder-staging.2i2c.cloud/ Image build and push succeeded, but spawn failed with:

```bash
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"pods \"jupyter-binder-2dexamples-2drequirements-2dat9cvbbg\" is forbidden: error looking up service account binder-staging/user-sa: serviceaccount \"user-sa\" not found","reason":"Forbidden","details":{"name":"jupyter-binder-2dexamples-2drequirements-2dat9cvbbg","kind":"pods"},"code":403}


Spawn failed: (403)
Reason: error
HTTP response headers: HTTPHeaderDict({'Audit-Id': '9b2b8bc3-ab2b-4fed-8d75-0fcacf2b86a4', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Kubernetes-Pf-Flowschema-Uid': '6076e3d8-fb9c-4af6-bc16-1828eab8bf98', 'X-Kubernetes-Pf-Prioritylevel-Uid': '3b5efc35-8cfb-4eeb-8300-d32a394811b8', 'Date': 'Wed, 08 Jun 2022 16:33:06 GMT', 'Content-Length': '369'})
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"pods \"jupyter-binder-2dexamples-2drequirements-2dat9cvbbg\" is forbidden: error looking up service account binder-staging/user-sa: serviceaccount \"user-sa\" not found","reason":"Forbidden","details":{"name":"jupyter-binder-2dexamples-2drequirements-2dat9cvbbg","kind":"pods"},"code":403}
```

**Comment by sgibson91** [comment]
Remove `singleuser.serviceAccountName` to remove the above issue

**Comment by sgibson91** [comment]
### Current problems

* Dask workers are not scaling up
  * I have been testing using this repo: https://github.com/TomAugspurger/pangeo-dask-gateway
  * Launch using this link: https://dask-staging.2i2c.cloud/hub/user-redirect/git-pull?repo=https%3A%2F%2Fgithub.com%2FTomAugspurger%2Fpangeo-dask-gateway&urlpath=tree%2Fpangeo-dask-gateway%2F&branch=master
  * However, a dask cluster is never successfully created. I can't remember if we need a service account or not for that to work - can anyone remind me?
* Validation failure
  * I am not sure how to resolve this:
    ```
    Error: execution error at (binderhub/charts/dask-gateway/templates/gateway/secret.yaml:10:27): gateway.auth.jupyterhub.apiToken must be defined when using jupyterhub auth
    ```

**Comment by GeorgianaElena** [comment]
> Error: execution error at (binderhub/charts/dask-gateway/templates/gateway/secret.yaml:10:27): gateway.auth.jupyterhub.apiToken must be defined when using jupyterhub auth

I believe that until https://github.com/dask/dask-gateway/issues/473 is solved, a token must be manually generated and explicitly set for the jupyterhub gateway service. The code here probably makes more sense than what I'm saying.

https://github.com/2i2c-org/infrastructure/blob/7b6d316959e60106bf4e90e2392a2111d225ff16/deployer/hub.py#L169-L186 
https://github.com/2i2c-org/infrastructure/blob/7b6d316959e60106bf4e90e2392a2111d225ff16/deployer/hub.py#L177-L186

**Comment by sgibson91** [comment]
> I believe that until [dask/dask-gateway#473](https://github.com/dask/dask-gateway/issues/473) is solved, a token must be manually generated and explicitly set for the jupyterhub gateway service.

So do we manually set this for other dask hubs? I feel like we don't, we let the deployer generate one. Therefore, if we wait for the deployer to generate one, how do those hubs pass the validation check - because the validation step runs before the deployer generates and inserts the token?

Am I misunderstanding something?

**Comment by GeorgianaElena** [comment]
@sgibson91, validation works for other daskhubs because of this (had no idea about it and never noticed it)...

https://github.com/2i2c-org/infrastructure/blob/7b6d316959e60106bf4e90e2392a2111d225ff16/deployer/config_validation.py#L113

File: helm-charts/binderhub/values.yaml [hunk 1]
```diff
@@ -11,31 +17,44 @@ binderhub:
...
+      emptyDir: {}
+  extraVolumeMounts:
+    - name: custom-templates
+      mountPath: /etc/binderhub/custom
+  dind:
+    enabled: true
+    daemonset:
+      image:
+        name: docker
+        tag: 19.03.5-dind
```
Comment (yuvipanda) [inline helm-charts/binderhub/values.yaml:47]: We can just inherit the upstream binderhub default here, which is much more recent.

Comment (sgibson91) [reply helm-charts/binderhub/values.yaml]: Done in [`0ffba14` (#1404)](https://github.com/2i2c-org/infrastructure/pull/1404/commits/0ffba14f2411f4d5e756a675bf8bffa0c1147e66)

**Comment by sgibson91** [comment]
Got a warning in the logs about calico

```
2022-06-22T09:11:06Z [Warning] Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "1b7ca330c53fd745eeae4e2c6d6145e4ec90293d784b615366dfa70466ece06b": stat /var/lib/calico/nodename: no such file or directory: check that the calico/node container is running and has mounted /var/lib/calico/
```

**Comment by sgibson91** [comment]
Other than https://github.com/2i2c-org/infrastructure/pull/1404/#issuecomment-1162851590, this is deployed and ready for review! 🎉

**Review by yuvipanda — APPROVED** [review]
\o/ Let's go! THANKS @sgibson91!

**Comment by yuvipanda** [comment]
The calico warning is harmless for the most part, means the pod got scheduled on the node before networkpolicy had a chance to start up.

**Pull request merged**